crud/common/sharding.lua

```diff
@@ -51,4 +75,72 @@ function sharding.tuple_set_and_return_bucket_id(tuple, space, specified_bucket_
     return bucket_id
 end
 
+-- Get sharding key (actually field names) for a certain space.
+-- Returns a table with field names or nil.
+function sharding.get_ddl_sharding_key(space_name, box_obj)
+    dev_checks('string', 'table')
+
+    if M.ddl_sharding_keys_cache == nil then
+        M.ddl_sharding_keys_cache = utils.fetch_ddl_sharding_key(box_obj)
+    end
+
+    return M.ddl_sharding_keys_cache[space_name]
+end
+
+-- Make sure fields used in sharding key are present in primary index.
+-- TODO: invalidate table sharding_key_in_primary_index
+function sharding.is_sharding_key_in_primary_index(space_name, primary_index, fieldnos)
+    dev_checks('string', 'table', 'table')
+    if sharding_key_in_primary_index[space_name] == nil then
+        local primary_index_fieldno = utils.get_index_fieldno_map(primary_index)
+        for _, fieldno in ipairs(fieldnos) do
+            if primary_index_fieldno[fieldno] == false then
+                sharding_key_in_primary_index[space_name] = false
+                break
+            end
+        end
+    end
+
+    return sharding_key_in_primary_index[space_name]
+end
+
+-- Build an array with sharding key values.
+-- Returns a table with sharding keys values or nil.
+local function build_sharding_key(key, index_parts, sharding_key_fieldno_map)
+    dev_checks('table', 'table', 'table')
+    local sharding_key = {}
+    for i, k in ipairs(key) do
+        local fieldno = index_parts[i].fieldno
+        if sharding_key_fieldno_map[fieldno] == true then
+            table.insert(sharding_key, k)
+        end
+    end
+
+    return sharding_key
+end
+
+-- Build sharding key using DDL sharding key.
+-- Returns a table with sharding key or pair of nil with error.
+function sharding.build_ddl_sharding_key(space_obj, sharding_key, ddl_sharding_key, bucket_id)
+    dev_checks('table', '?', 'table', 'number|nil')
```

Comment:
Why not `?number` instead of `number|nil`?

Reply:
Because it accepts `opts.bucket_id` passed from outside by use that can be `nil`.